squidpy/gr/_ppatterns.py

```diff
@@ -196,75 +198,109 @@ def moran(
             genes = adata.var_names.values
     genes = _assert_non_empty_sequence(genes, name="genes")
 
+    params: Dict[str, Any] = {"mode": mode, "transformation": transformation, "two_tailed": two_tailed}
+
+    if params["mode"] == "moran":
+        params["func"] = _morans_i
+        params["stat"] = "I"
+        params["expected"] = -1.0 / (adata.shape[0] - 1)  # expected score
+    elif params["mode"] == "geary":
+        params["func"] = _gearys_c
+        params["stat"] = "C"
+        params["expected"] = 1.0
+    else:
+        raise NotImplementedError(f"mode: `{mode}` is not available.")
+
     n_jobs = _get_n_cores(n_jobs)
-    start = logg.info(f"Calculating for `{len(genes)}` genes using `{n_jobs}` core(s)")
-
-    w = _set_weight_class(adata, key=connectivity_key)  # init weights
-    df = parallelize(
-        _moran_helper,
-        collection=genes,
-        extractor=pd.concat,
-        use_ixs=True,
-        n_jobs=n_jobs,
-        backend=backend,
-        show_progress_bar=show_progress_bar,
-    )(adata=adata, weights=w, transformation=transformation, permutations=n_perms, layer=layer, seed=seed)
+
+    vals = _get_obs_rep(adata[:, genes], use_raw=use_raw, layer=layer).T
+    g = adata.obsp[connectivity_key].copy()
+    # row-normalize
+    if transformation:
+        normalize(g, norm="l1", axis=1, copy=False)
+
+    res = np.empty((vals.shape[0], 6), dtype=np.float32) * np.nan
+
+    score = params["func"](
+        g,
+        vals,
+    )
+
+    start = logg.info(f"Calculating for `{n_perms}` permutations using `{n_jobs}` core(s)")
```

Comment:
I'd add which statistic we're computing: `Calculating {mode}'s statistic for ...` or similar.